sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/LongLivedMessageSubscriber.java

```diff
@@ -0,0 +1,168 @@
+
+        currentWork.next(message);
+
+        logger.verbose("[{}] received message with Sequence Number [{}].", currentWork.getId(), message.getMessage().getSequenceNumber());
+
+        if (currentWork.isTerminal()) {
+            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", currentWork.getId());
+            completeCurrentWork(currentWork);
+        }
+    }
+
+    private void completeCurrentWork(SynchronousReceiveWork currentWork) {
+
+        if (isTerminated()) {
+            return;
+        }
+
+        currentWork.complete();
+        logger.verbose("[{}] work completed.", currentWork.getId());
+
+        if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
+            timeoutOperation.dispose();
+        }
+        if (drainQueueDisposable != null && !drainQueueDisposable.isDisposed()) {
+            drainQueueDisposable.dispose();
+        }
+
+        if (wip.decrementAndGet() != 0) {
+            logger.warning("There is another worker in drainLoop. But there should only be 1 worker. Value:"+wip.get());
+        }
+
+        // After current work finished and there more receive requests
+        if (workQueue.size() > 0 ) {
+            drain();
+        }
+    }
+
+    void queueWork(SynchronousReceiveWork work) {
+        workQueue.add(work);
+        drain();
+    }
+
+    private void drain() {
+        // If someone is already in this loop, then we are already clearing the queue.
+        if (!wip.compareAndSet(0, 1)) {
+            return;
+        }
+        // Drain queue..
+        drainQueueDisposable = Mono.just(true)
+            .subscribe(l -> {
+                drainQueue();
+            });
+    }
+
+    private void drainQueue() {
+        if (isTerminated()) {
+            return;
+        }
+        currentWork = workQueue.poll();
+        if (currentWork == null) {
+            return;
+        }
+        subscription.request(currentWork.getNumberOfEvents());
+
+        // timer to complete the current in case of timeout trigger
+        timeoutOperation = Mono.delay(currentWork.getTimeout())
+            .subscribe(l -> {
+                if (!currentWork.isTerminal()) {
+                    completeCurrentWork(currentWork);
+                }
+            });
+    }
+
+
+    @Override
+    protected void hookOnComplete() {
+        logger.info("[{}] Completed. No events to listen to.", currentWork.getId());
+        dispose();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void hookOnError(Throwable throwable) {
+        logger.error("[{}] Errors occurred upstream", currentWork.getId(), throwable);
+        currentWork.error(throwable);
+        dispose();
+    }
+
+    @Override
+    protected void hookOnCancel() {
+        dispose();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void dispose() {
+        if (isDisposed.getAndSet(true)) {
```

Comment:
We don't need to override dispose. This is done in the basesubscriber. We can hook into `hookOnCancel()`.